### PR TITLE
Added distributed LDG method

### PR DESCRIPTION
**Pull request by mdh266**

### CODE REVIEW

File: Distributed_LDG_Method/README.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
...ethods. While this is very convenient, I could not use this framework for solving my problem and I needed write the LDG method from scratch. I thought it would help for others to have access to this example that goes through writing a discontinuous Galerkin method from scatch and also shows how to do it in a distributed setting using the <a href="https://www.trilinos.org">Trilinos</a> library. This example may also be of interest to users that wish to use the LDG method, as the method is distinctly different from the Interior Penalty Discontinuous Galerkin (IPDG) methods and was not covered in other tutorials on DG methods.  The LDG method is very useful when one is working with a differential equation and desires both approximations to the scalar unknown function as well as its flux.  The mixed method is another method where one can obtain both the scalar unknown function as we as its flux, however, the LDG method has less degrees of freedom when using the Raviart-Thomas elements in the mixed finite element. It also approximates the scalar unknown function and its flux using discontinous polynomial basis functions and are much more suitable when one wishes to use local refinement.
```
Comment (bangerth) [inline Distributed_LDG_Method/README.md]: as we as -> as well as

(if you break the lines, it's going to be easier to annotate :-)

Comment (bangerth) [inline Distributed_LDG_Method/README.md]: "the LDG method has less degrees of freedom when using the Raviart-Thomas elements"

Did you mean to say "has less degrees of freedom COMPARED to using..."? Also, elements -> element

Comment (bangerth) [inline Distributed_LDG_Method/README.md]: discontinous -> discontinuous

File: Distributed_LDG_Method/doc/author [hunk 2]
```diff
@@ -0,0 +1 @@
+Michael Harmon <mdh266@gmail.com>
```
Comment (bangerth) [inline Distributed_LDG_Method/doc/author]: add a newline to the end of this file. same for builds-on, dependencies, entry-name

File: Distributed_LDG_Method/doc/tooltip [hunk 3]
```diff
@@ -0,0 +1 @@
+Solving Poisson's Equation Using The Local Discontinuous Galerkin Methods And Trilinos
```
Comment (bangerth) [inline Distributed_LDG_Method/doc/tooltip]: and here

I think the other tooltips don't capitalize all words -- maybe go with that style?

**Comment by mdh266** [comment]
Thanks! Do I need to make sure all the lines in the cc file are within 80 characters as well?

**Comment by bangerth** [comment]
No. I personally find that helpful, but we don't review the code for code gallery projects.

**Comment by mdh266** [comment]
Okay, thanks, I added the changes and tried to keep everything to around 80 characters in the code as well.

File: Distributed_LDG_Method/README.md [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+and,
+@f{align}
+\{\textbf{f} \} \; = \; \frac{1}{2}(\textbf{f}^- + \textbf{f}^+), 
+\qquad \mbox{and}\qquad  
+\left[ \textbf{f} \right] \; = \;\textbf{f}^+ \cdot \textbf{n}^+ + 
+\textbf{f}^- \cdot \textbf{n}^- , 
+@f}
+
+where $f$ is a scalar function and $\textbf{f}$ is vector-valued function. 
+We note that for a faces that are on the boundary of the domain we have,
```
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: "for faces"

File: Distributed_LDG_Method/README.md [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+\qquad \mbox{and}\qquad  
+\left[ \textbf{f} \right] \; = \; \textbf{f} \cdot \textbf{n}.
+@f}
+
+
+We denote the volume integrals and surface integrals using the $L^{2}(\Omega)$
+inner products by $( \, \cdot \, , \, \cdot \, )_{\Omega}$ and $\langle  \, 
+\cdot \, , \, \cdot \,  \rangle_{\partial \Omega}$ respectively. 
+
+As with the mixed finite element method, the LDG discretization requires the
```
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: Again, "the mixed FE method" refers to a specific implementation. Can you please clarify this? Maybe you just need make reference to the use of R-T elements again.

File: Distributed_LDG_Method/README.md [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+would be helpful for others to have access to 
+this example that goes through writing a discontinuous Galerkin method from 
+scatch and also shows how to do it in a distributed setting using the 
+<a href="https://www.trilinos.org">Trilinos</a> library. This example may also
+be of interest to users that wish to use the LDG method, as the method is 
+distinctly different from the Interior Penalty Discontinuous Galerkin (IPDG) 
+methods and was not covered in other tutorials on DG methods.  The LDG method
+is very useful when one is working with a differential equation and desires 
+both approximations to the scalar unknown function as well as its flux.  The 
+mixed method is another method where one can obtain both the scalar unknown
```
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: > The mixed method

Could you please clarify this a little, perhaps by referring to a specific method? I understand [mixed methods](https://en.wikipedia.org/wiki/Mixed_finite_element_method) to be an entire category of approaches rather than a single, unique method (even if there is only one that's regularly applied to this problem class). 

Alternatively, you could make this statement more generic, e.g.

> The application of a mixed method offers a mechanism whereby one can...

File: Distributed_LDG_Method/README.md [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+
+## Local Discontinuous Galerkin Method
+
+In this section we discuss the LDG method and first introduce some notation. 
+Let $\mathcal{T}_{h} = \mathcal{T}_{h}(\Omega) \, = \, \left\{ \, \Omega_{e} 
+\, \right\}_{e=1}^{N}$ be  the general triangulation of a domain $\Omega \; 
+\subset \; \mathbb{R}^{d}, \; d \, = \, 1, 2, 3$, into $N$ non-overlapping 
+elements $\Omega_{e}$ of diameter $h_{e}$.  The maximum size of the diameters
+of all the elements is $h = \max( \, h_{e}\, )$.  We define $\mathcal{E}_{h}$
```
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: "of all elements"

File: Distributed_LDG_Method/README.md [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+\textbf{n} & \ \text{in} \ \mathcal{E}_{D} \\
+\end{array}
+\right.
+@f}
+
+
+The term $\boldsymbol \beta$ is a constant unit vector which does not lie 
+parallel to any element face in $ \mathcal{E}_{h}^{i}$.  For 
+$\boldsymbol \beta =  0$,  $\widehat{\textbf{q}_{h}}$ and $\widehat{u_{h}}$ 
+are called the central or Brezzi et. al. fluxes. For
```
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: Do you have a citation (e.g. a DOI link) for "Brezzi et al."?

File: Distributed_LDG_Method/README.md (hunk 6 above)
Comment (mdh266) [inline Distributed_LDG_Method/README.md]: Sure, is this what you meant by the suggestion: 

The LDG method is very useful when one is working with a differential equation and desires 
both approximations to the scalar unknown function as well as its flux. 
The application of a mixed method offers a mechanism whereby one can
can obtain both the scalar unknown function as well as its flux, however, 
the LDG method has less degrees of freedom compared to the mixed method with 
the Raviart-Thomas element.

Comment (mdh266) [inline Distributed_LDG_Method/README.md]: Or did you mean:

The application of a mixed method offers a mechanism whereby one can
can obtain both the scalar unknown function as well as its flux, however, in
the case of Poisson's equation the LDG method has less degrees of freedom compared to the mixed method with the Raviart-Thomas element.

File: Distributed_LDG_Method/README.md (hunk 8 above)
Comment (mdh266) [inline Distributed_LDG_Method/README.md]: Thanks! I made the changes you suggested, although the image wouldnt show up on my local build of the documentation.

File: Distributed_LDG_Method/README.md (hunk 6 above)
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: Hi @mdh266! IMHO your first suggestion not only satisfies my inquiry, but also offers a more complete motivation for the use of the LDG method. So I'd be happy if you used that one :-)

File: Distributed_LDG_Method/README.md (hunk 8 above)
Comment (jppelteret) [inline Distributed_LDG_Method/README.md]: Cool, thanks for considering my suggestions :-)

File: Distributed_LDG_Method/README.md (hunk 6 above)
Comment (mdh266) [inline Distributed_LDG_Method/README.md]: Cool, thanks!

Comment (bangerth) [inline Distributed_LDG_Method/README.md]: has less degrees of freedom -> has FEWER degrees of freedom :-)

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 9]
```diff
@@ -0,0 +1,1514 @@
...
+    locally_relevant_solution.reinit(locally_relevant_dofs,
+                                     MPI_COMM_WORLD);
+
+    system_rhs.reinit(locally_owned_dofs,
+                      locally_relevant_dofs,
+                      MPI_COMM_WORLD,
+                      true);
+
+    const unsigned int n_vector_field = dim * dofs_per_component[0];
+    const unsigned int n_potential = dofs_per_component[1];
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: use `dofs_per_component[dim]` for consistency.

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 10]
```diff
@@ -0,0 +1,1514 @@
...
+
+    // We first remark that we have the FEValues objects for 
+    // the values of our cell basis functions as was done in most
+    // other examples.  Now because we 
+    // are using discontinuous Galerkin methods we also introduce a 
+    // FEFaceValues object, <code>fe_face_values</code>, 
+    // for evaluating the basis functions
+    // on one side of an element face as well as another FEFaceValues object, 
+    // <code>fe_neighbor_face_values</code>, for evaluting the basis functions 
+    // on the opposite side of the face, i.e. on the neighoring element's face.
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: neighboring

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 11]
```diff
@@ -0,0 +1,1514 @@
...
+                                local_neighbor_dof_indices(dofs_per_cell);                                      
+
+    // We first remark that we have the FEValues objects for 
+    // the values of our cell basis functions as was done in most
+    // other examples.  Now because we 
+    // are using discontinuous Galerkin methods we also introduce a 
+    // FEFaceValues object, <code>fe_face_values</code>, 
+    // for evaluating the basis functions
+    // on one side of an element face as well as another FEFaceValues object, 
+    // <code>fe_neighbor_face_values</code>, for evaluting the basis functions
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: evaluating

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 12]
```diff
@@ -0,0 +1,1514 @@
...
+    // are using discontinuous Galerkin methods we also introduce a 
+    // FEFaceValues object, <code>fe_face_values</code>, 
+    // for evaluating the basis functions
+    // on one side of an element face as well as another FEFaceValues object, 
+    // <code>fe_neighbor_face_values</code>, for evaluting the basis functions 
+    // on the opposite side of the face, i.e. on the neighoring element's face. 
+    // In addition, we also introduce a FESubfaceValues object,
+    // <code>fe_subface_values</code>, that
+    // will be used for dealing with faces that have multiple refinement 
+    // levels, i.e. hanging nodes. When we have to evaulate the fluxes across
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: evaluate

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 13]
```diff
@@ -0,0 +1,1514 @@
...
+
+    // The next four matrices are used to incorporate the flux integrals across
+    // interior faces of the mesh:
+    FullMatrix<double>      vi_ui_matrix(dofs_per_cell, dofs_per_cell);
+    FullMatrix<double>      vi_ue_matrix(dofs_per_cell, dofs_per_cell);
+    FullMatrix<double>      ve_ui_matrix(dofs_per_cell, dofs_per_cell);
+    FullMatrix<double>      ve_ue_matrix(dofs_per_cell, dofs_per_cell);
+    // As explained in the section on the LDG method we take our test
+    // function to be v and multiply it on the left side of our differential 
+    // equation that is on u and peform integration by parts as explain in the
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: explained

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 14]
```diff
@@ -0,0 +1,1514 @@
...
+                                local_matrix,
+                                local_vector);
+
+            // We remark that we need to get the local indices for the dofs to
+            // to this cell before we begin to compute the contributions 
+            // from the numerical fluxes, i.e. the boundary conditions and
+            // interior fluxes.
+            cell->get_dof_indices(local_dof_indices);
+
+            // Now is where were start to loop over all the faces of the cell
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: were -> we (?)

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 15]
```diff
@@ -0,0 +1,1514 @@
...
+                        // the work to assemble the interior flux matrices 
+                        // is very much the same as before. Infact it is
+                        // much simpler since we do not have to loop through the
+                        // subfaces.  However, we do have to check that we do
+                        // not compute the same contribution twice. Since we are
+                        // looping over all the faces of all the cells in the mesh,
+                        // we pass over each face twice.  If we do not take this 
+                        // into consideration when assembling the interior flux 
+                        // matrices we might compute the local interior flux matrix
+                        // twice. To avoid doing this we only compute the interior
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: You use "twice" three times in a row. You might want to make this a bit more compact.

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 16]
```diff
@@ -0,0 +1,1514 @@
...
+                        // looping over all the faces of all the cells in the mesh,
+                        // we pass over each face twice.  If we do not take this 
+                        // into consideration when assembling the interior flux 
+                        // matrices we might compute the local interior flux matrix
+                        // twice. To avoid doing this we only compute the interior 
+                        // fluxes once for each face by restricting that the 
+                        // following computation only occur on the on 
+                        // the cell face with the lower index number.
+                        if(neighbor->level() == cell->level() &&
+                            neighbor->index() > cell->index() )
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: Have you verified that this is also stable when done in parallel with MPI? I'm not sure whether the index of the cells is really laid out the same way on all processors... I would recommend using `neighbor->id() < cell->id()` because that is stable also with distributed triangulations. However, when you use the `CellId` you will also need a similar switch in the hanging node put further up.

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 17]
```diff
@@ -0,0 +1,1514 @@
...
+                            // matrices to the global system matrix.
+                            // This is done through the class function
+                            // <code>distribute_local_flux_to_global()</code>.
+                            distribute_local_flux_to_global(
+                                                      vi_ui_matrix,
+                                                      vi_ue_matrix,
+                                                      ve_ui_matrix,
+                                                      ve_ue_matrix,
+                                                      local_dof_indices,
+                                                      local_neighbor_dof_indices);
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: Could you please add a note why you don't use `local_matrix` instead of `vi_ui_matrix`? It makes sense because you'd need to skip that matrix in the `distribute_local_flux_to_global` but it might be good to explain it anyway.

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 18]
```diff
@@ -0,0 +1,1514 @@
...
+} 
+
+// @sect4{distribute_local_flux_to_global}
+// In this function we use the ConstraintMatrix to distribute
+// the local flux matrices to the global system matrix.  
+// Since I have to do this twice in assembling the 
+// system matrix, I made function to do it rather than have
+// repeated code.
+// We remark that the reader take special note of
+// the which matrices we are distributing and the order
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: remove "the"

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 19]
```diff
@@ -0,0 +1,1514 @@
...
+    // to solution vector which has some overlap between 
+    // processors, <code>locally_relevant_solution</code>.
+    // We need the overlapped portions of our solution
+    // in order to be able to do computations using the solution
+    // later in the code or in post processing.
+    locally_relevant_solution = completely_distributed_solution;
+}
+
+// @sect4{output_results}
+// This functio
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: results

Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: won't

**Comment by kronbichler** [comment]
One more comment: Would you mind running the deal.II indentation script on the code? Not that we force it in gallery projects but I find it useful to have a uniform style. This would be given by `astyle --options=contrib/styles/astyle.rc LDPoisson.cc` with astyle 2.04.

File: Distributed_LDG_Method/LDGPoisson.cc [hunk 20]
```diff
@@ -0,0 +1,1514 @@
...
+// the Trilinos library.
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/trilinos_vector.h>
+#include <deal.II/lac/trilinos_solver.h>
+
+
+// The functions class contains all the defintions of the functions we 
+// will use, i.e. the right hand side function, the boundary conditions 
+// and the test functions.  
+#include "Functions.cc"
```
Comment (kronbichler) [inline Distributed_LDG_Method/LDGPoisson.cc]: I wouldn't call this `Functions.cc` but `Functions.h` because you do not instantiate anything in the `Functions.cc` file. Consequently, you could also remove it from `CMakeLists.txt`.

File: Distributed_LDG_Method/LDGPoisson.cc (hunk 9 above)
Comment (mdh266) [inline Distributed_LDG_Method/LDGPoisson.cc]: good catch, thanks!

File: Distributed_LDG_Method/LDGPoisson.cc (hunk 20 above)
Comment (mdh266) [inline Distributed_LDG_Method/LDGPoisson.cc]: The boundary conditions are in Functions.cc and  in the constructor of LDGPoisson class I instantiate them, does this not count?

File: Distributed_LDG_Method/LDGPoisson.cc (hunk 17 above)
Comment (mdh266) [inline Distributed_LDG_Method/LDGPoisson.cc]: Do you mean why we have have to introduce flux matrices or why we dont distribute the local_matrix at this point as well?

File: Distributed_LDG_Method/LDGPoisson.cc (hunk 16 above)
Comment (mdh266) [inline Distributed_LDG_Method/LDGPoisson.cc]: I switched it for this case you mentioned, but I'm not sure what you mean by dealing with the hanging nodes?  I never used the index of the cells in dealing with the case of a hanging node.

**Comment by mdh266** [comment]
Is there anything else I should change or update on this comit?

**Comment by bangerth** [comment]
Let's merge it and see how it turns out!

**Comment by mdh266** [comment]
I see on the code gallery i made a few typos with the latex, mainly i need to replace \rrbackets with ] and  the image also didnt come out well.. is including the image not the usual markdown command name ?

should i make another pull request when done?

**Comment by bangerth** [comment]
Yes, create a separate pull request that fixes issues you see.

I don't know enough about markdown -- but you can play with different versions and create the HTML pages of the code gallery locally until you find that it works.
